### PR TITLE
requestTypes design doc

### PR DESCRIPTION
**Pull request by Prateeknandle**
Signed-off-by: Prateeknandle <prateeknandle@gmail.com>

### CODE REVIEW

File: proposals/add_requestTypes_to_match_&_exclude.md [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+
+- [Kyverno Design Proposal - Adding Request Types to match and exclude](#kyverno-design-proposal---adding-request-types-to-match-and-exclude)
+  * [Introduction](#introduction)
+  * [Proposal](#proposal)
+  * [Example](#example)
+  * [Implementation](#implementation)
+
+## Introduction
+
+Until now we use request.operation under condition.key, to filter a request, now we'll use a tag requestTypes in match or exclude blocks to match specific requests comming from admission review requests, & then according to the defined conditions the policy would apply on to the resource.
```
Comment (realshuting) [inline proposals/add_requestTypes_to_match_&_exclude.md:17]: Typo: "coming"

File: proposals/add_requestTypes_to_match_&_exclude.md [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+        kinds:
+        - Pod
+    validate:
+      message: "label 'app.kubernetes.io/name' is required"
+      deny:
+        conditions:
+          any:
+          - key: "{{request.operation}}"
+            operator: Equals
+            value: DELETE
```
Comment (realshuting) [inline proposals/add_requestTypes_to_match_&_exclude.md:48]: The feature is to add request operations as built-in variables, this validate policy defines nothing else but to reject DELETE requests. Please update this example to reflect some other use case.

File: proposals/add_requestTypes_to_match_&_exclude.md [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+      pattern:
+        metadata:
+          labels:
+            app.kubernetes.io/name: "?*"
+```
+
+When a pod with label `app.kubernetes.io/name: "?*"` is requested to be deleted, the request will match with the requestTypes, but the validate.deny condition will block the delete operation.
```
Comment (realshuting) [inline proposals/add_requestTypes_to_match_&_exclude.md:55]: Where do we match this label in the above policy?

File: proposals/add_requestTypes_to_match_&_exclude.md [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+      resources:
+        kinds:
+        - Pod
+    validate:
+      message: "label 'app.kubernetes.io/name' is required"
+        conditions:
+          any:
+          - key: "{{request.operation}}"
+            operator: Equals
+            value: DELETE
```
Comment (realshuting) [inline proposals/add_requestTypes_to_match_&_exclude.md:79]: Why do we need this condition?

**Review by realshuting — CHANGES_REQUESTED** [review]
@Prateeknandle - I left a couple of comments below. It seems that there's some misunderstanding about this feature. 

I would suggest reading through this [doc](https://kyverno.io/docs/writing-policies/variables/#pre-defined-variables) to understand how built-in variables work currently, and re-visit this design.

**Comment by Prateeknandle** [comment]
> I would suggest reading through this [doc](https://kyverno.io/docs/writing-policies/variables/#pre-defined-variables) to understand how built-in variables work currently, and re-visit this design.

@realshuting we don't need to assign the value to `request.operation` as shown in [here](https://kyverno.io/docs/writing-policies/variables/#pre-defined-variables), for different variables, because kubernetes-api will assign request-type, value to the variable `request.operation`.
So basically we need to access the `request.operation` and then use it's value to match with the new tag `requestTypes`.
Is this correct?

**Comment by Prateeknandle** [comment]
Need to close this pr, actually I pushed it from main branch, reopening new pr. 
Sorry for the inconvenience

**Pull request closed without merging**